web/src/engine/osk/src/input/gestures/browser/keytip.ts

```diff
@@ -139,32 +167,56 @@ export default class KeyTip implements KeyTipInterface {
       this.cap.style.width = xWidth + 'px';
       this.tip.style.height = halfHeight + 'px';
 
-      this.cap.style.top = (halfHeight - 3) + 'px';
-      this.cap.style.height = (keyRect.bottom - _BoxRect.top - Math.floor(y - canvasHeight) - (halfHeight)) + 'px'; //(halfHeight + 3 + ySubPixelPadding) + 'px';
+      const capOffset = 3;
+      if(orientation == 'up') {
+        this.cap.style.top = (halfHeight - capOffset) + 'px';
+        this.cap.style.bottom = '';
+      } else {
+        this.cap.style.top = '';
+        this.cap.style.bottom = (halfHeight - capOffset) + 'px';
+      }
+      const defaultCapHeight = (distFromTop - Math.floor(y) + canvasHeight - (orientation == 'up' ? halfHeight : -capOffset * 2));
+      this.cap.style.height = defaultCapHeight + 'px';
 
+      // let delta: number = 0;
       if(this.constrain && tipHeight + bottomY > oskHeight) {
         const delta = tipHeight + bottomY - oskHeight;
         kts.height = (canvasHeight-delta) + 'px';
         const hx = Math.max(0, (canvasHeight-delta)-(canvasHeight/2) + 2);
         this.cap.style.height = hx + 'px';
+      } else if(bottomY < 0) { // we'll assume that we always constrain at the OSK's bottom.
+        kts.bottom = '0px';
+        this.cap.style.height = (defaultCapHeight + bottomY) + 'px';
```

Comment:
Not 100% sure if it's necessary, but it can't hurt to have:

```suggestion
        this.cap.style.height = Math.max(0, defaultCapHeight + bottomY) + 'px';
```

Don't want to leave a chance of the 'cap' covering the preview should there be a valid flick key in the final keyboard row for some reason.